fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java

```diff
@@ -718,6 +723,68 @@ protected RuntimeException mapAsyncToSyncException(@Nonnull Throwable ex) {
         }
     }
 
+    /**
+     * Join a future, following the same logic that asyncToSync uses to validate that the operation isn't blocking
+     * in an asynchronous context.
+     *
+     * @param future the future to be completed
+     * @param <T> the type of the value produced by the future
+     * @return the result value
+     */
+    public <T> T join(CompletableFuture<T> future) {
+        checkIfBlockingInAsync(future);
+        return future.join();
+    }
+
+    /**
+     * Get a future, following the same logic that asyncToSync uses to validate that the operation isn't blocking
+     * in an asynchronous context.
+     *
+     * @param future the future to be completed
+     * @param <T> the type of the value produced by the future
+     * @return the result value
+     *
+     * @throws java.util.concurrent.CancellationException if the future was cancelled
+     * @throws ExecutionException if the future completed exceptionally
+     * @throws InterruptedException if the current thread was interrupted
+     */
+    public <T> T get(CompletableFuture<T> future) throws InterruptedException, ExecutionException {
+        checkIfBlockingInAsync(future);
+        return future.get();
+    }
+
+    @API(API.Status.INTERNAL)
+    public BlockingInAsyncDetection getBlockingInAsyncDetection() {
+        return blockingInAsyncDetectionSupplier.get();
+    }
+
+    private void checkIfBlockingInAsync(CompletableFuture<?> future) {
+        BlockingInAsyncDetection behavior = getBlockingInAsyncDetection();
+        if (behavior == BlockingInAsyncDetection.DISABLED) {
+            return;
+        }
+
+        final boolean isComplete = future.isDone();
+        if (isComplete && behavior.ignoreComplete()) {
+            return;
+        }
+
+        final StackTraceElement[] stack = Thread.currentThread().getStackTrace();
+        for (StackTraceElement stackElement : stack) {
+            if (stackElement.getClassName().startsWith(CompletableFuture.class.getName())) {
+                final Exception locationException = new Exception("Location of blocking call");
```

Comment:
Wait, I had an idea about this. Could this be a `BlockingInAsyncException`? Then if it's supposed to warn, it's the exception that gets logged; if it's supposed to throw, it's the thing that gets thrown. It's probably six of one, half dozen of the other, but it might be a little cleaner.